File: src/main/java/work/lclpnet/ap2/impl/game/Announcer.java
package work.lclpnet.ap2.impl.game;

import net.fabricmc.fabric.api.networking.v1.PlayerLookup;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.sound.SoundCategory;
import net.minecraft.sound.SoundEvent;
import net.minecraft.sound.SoundEvents;
import net.minecraft.text.Text;
import org.jetbrains.annotations.Nullable;
import work.lclpnet.kibu.title.Title;
import work.lclpnet.kibu.translate.TranslationService;
import work.lclpnet.kibu.translate.text.TranslatedText;

import static net.minecraft.util.Formatting.AQUA;
import static net.minecraft.util.Formatting.DARK_GREEN;

public class Announcer {

    private final TranslationService translations;
    private final MinecraftServer server;
    @Nullable
    private SoundEvent sound = SoundEvents.BLOCK_NOTE_BLOCK_PLING.value();
    private SoundCategory category = SoundCategory.RECORDS;
    private float volume = 0.5f;
    private float pitch = 0.5f;
    private int fadeInTicks = 5;
    private int stayTicks = 30;
    private int fadeOutTicks = 5;

    public Announcer(TranslationService translations, MinecraftServer server) {
        this.translations = translations;
        this.server = server;
    }

    public Announcer withDefaults() {
        this.sound = SoundEvents.BLOCK_NOTE_BLOCK_PLING.value();
        this.category = SoundCategory.RECORDS;
        this.volume = 0.5f;
        this.pitch = 0.5f;
        this.fadeInTicks = 5;
        this.stayTicks = 30;
        this.fadeOutTicks = 5;
        return this;
    }

    public Announcer silent() {
        this.sound = null;
        return this;
    }

    public Announcer withSound(@Nullable SoundEvent sound, SoundCategory category, float volume, float pitch) {
        this.sound = sound;
        this.category = category == null ? SoundCategory.NEUTRAL : category;
        this.volume = Math.max(0f, volume);
        this.pitch = Math.max(0.5f, Math.min(2f, pitch));
        return this;
    }

    public Announcer withTimes(int fadeInTicks, int stayTicks, int fadeOutTicks) {
        this.fadeInTicks = Math.max(0, fadeInTicks);
        this.stayTicks = Math.max(0, stayTicks);
        this.fadeOutTicks = Math.max(0, fadeOutTicks);
        return this;
    }

    public void announceSubtitle(String translationKey) {
        announce(null, translationKey);
    }

    public void announce(@Nullable String titleKey, @Nullable String subtitleKey) {
        TranslatedText title = titleKey != null ? translations.translateText(titleKey).formatted(AQUA) : null;
        TranslatedText subtitle = subtitleKey != null ? translations.translateText(subtitleKey).formatted(DARK_GREEN) : null;

        announce(title, subtitle);
    }

    public void announce(@Nullable TranslatedText title, @Nullable TranslatedText subtitle) {
        for (ServerPlayerEntity player : PlayerLookup.all(server)) {
            Text titleText = title != null ? title.translateFor(player) : Text.empty();
            Text subTitleText = subtitle != null ? subtitle.translateFor(player) : Text.empty();

            Title.get(player).title(titleText, subTitleText, fadeInTicks, stayTicks, fadeOutTicks);

            if (sound == null) continue;

            player.playSound(sound, category, volume, pitch);
        }
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/game/util/Announcer.kt
package work.lclpnet.ap2.game.util

import net.fabricmc.fabric.api.networking.v1.PlayerLookup
import net.minecraft.ChatFormatting
import net.minecraft.network.chat.Component
import net.minecraft.server.MinecraftServer
import net.minecraft.server.level.ServerPlayer
import net.minecraft.sounds.SoundEvent
import net.minecraft.sounds.SoundEvents
import net.minecraft.sounds.SoundSource
import work.lclpnet.ap2.ext.mc.playNotifySound
import work.lclpnet.ap2.ext.server
import work.lclpnet.ap2.ext.translations
import work.lclpnet.ap2.game.MiniGameInstance
import work.lclpnet.kibu.title.Title
import work.lclpnet.kibu.translate.Translations
import work.lclpnet.kibu.translate.text.TranslatedText

data class Announcer(
    private val translations: Translations,
    private val players: () -> Iterable<ServerPlayer>,
    private val sound: SoundEvent? = SoundEvents.NOTE_BLOCK_PLING.value(),
    private val category: SoundSource = SoundSource.RECORDS,
    private val volume: Float = 0.5f,
    private val pitch: Float = 0.5f,
    private val fadeInTicks: Int = 5,
    private val stayTicks: Int = 30,
    private val fadeOutTicks: Int = 5,
) {

    constructor(translations: Translations, server: MinecraftServer) : this(
        translations,
        { PlayerLookup.all(server) }
    )

    fun silent() = copy(sound = null)

    fun withSound(sound: SoundEvent, category: SoundSource, volume: Float, pitch: Float) = copy(
        sound = sound,
        category = category,
        volume = volume.coerceAtLeast(0f),
        pitch = pitch.coerceIn(0.5f..2f)
    )

    fun withTimes(fadeInTicks: Int, stayTicks: Int, fadeOutTicks: Int) = copy(
        fadeInTicks = fadeInTicks.coerceAtLeast(0),
        stayTicks = stayTicks.coerceAtLeast(0),
        fadeOutTicks = fadeOutTicks.coerceAtLeast(0),
    )

    fun announceSubtitle(translationKey: String) {
        announce(null, translationKey)
    }

    fun announce(titleKey: String?, subtitleKey: String?) {
        val title = if (titleKey != null) {
            translations.translateText(titleKey).withStyle(ChatFormatting.AQUA)
        } else null

        val subtitle = if (subtitleKey != null) {
            translations.translateText(subtitleKey).withStyle(ChatFormatting.DARK_GREEN)
        } else null

        announce(title, subtitle)
    }

    fun announce(title: TranslatedText?, subtitle: TranslatedText?) {
        for (player in players()) {
            val titleText = if (title != null) title.translateFor(player) else Component.empty()
            val subTitleText = if (subtitle != null) subtitle.translateFor(player) else Component.empty()

            Title.get(player).title(titleText, subTitleText, fadeInTicks, stayTicks, fadeOutTicks)

            sound?.let {
                player.playNotifySound(it, category, volume, pitch)
            }
        }
    }
}

fun MiniGameInstance.useAnnouncer() = Announcer(translations, server)